src/sort/sort.rs

```diff
@@ -41,22 +46,105 @@ enum SortMode {
 
 struct Settings {
     mode: SortMode,
+    merge: bool,
     reverse: bool,
     outfile: Option<String>,
     stable: bool,
     unique: bool,
     check: bool,
+    compare_fns: Vec<fn(&String, &String) -> Ordering>,
 }
 
 impl Default for Settings {
     fn default() -> Settings {
         Settings {
             mode: SortMode::Default,
+            merge: false,
             reverse: false,
             outfile: None,
             stable: false,
             unique: false,
             check: false,
+            compare_fns: Vec::new(),
+        }
+    }
+}
+
+struct MergeableFile<'a> {
+    lines: Lines<BufReader<Box<Read>>>,
+    current_line: String,
+    settings: &'a Settings,
+}
+
+// BinaryHeap depends on `Ord`. Note that we want to pop smallest items
+// from the heap first, and BinaryHeap.pop() returns the largest, so we
+// trick it into the right order by calling reverse() here.
+impl<'a> Ord for MergeableFile<'a> {
+    fn cmp(&self, other: &MergeableFile) -> Ordering {
+        compare_by(&self.current_line, &other.current_line, &self.settings).reverse()
+    }
+}
+
+impl<'a> PartialOrd for MergeableFile<'a> {
+    fn partial_cmp(&self, other: &MergeableFile) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<'a> PartialEq for MergeableFile<'a> {
+    fn eq(&self, other: &MergeableFile) -> bool {
+        Ordering::Equal == compare_by(&self.current_line, &other.current_line, &self.settings)
+    }
+}
+
+impl<'a> Eq for MergeableFile<'a> {}
+
+struct FileMerger<'a> {
+    heap: BinaryHeap<MergeableFile<'a>>,
+    settings: &'a Settings,
+}
+
+impl<'a> FileMerger<'a> {
+    fn new(settings: &'a Settings) -> FileMerger<'a> {
+        FileMerger {
+            heap: BinaryHeap::new(),
+            settings: settings,
+        }
+    }
+    fn push_file(&mut self, mut lines: Lines<BufReader<Box<Read>>>){
+        match lines.next() {
+            Some(Ok(next_line)) => {
+                let mergeable_file = MergeableFile {
+                    lines: lines,
+                    current_line: next_line,
+                    settings: &self.settings,
+                };
+                self.heap.push(mergeable_file);
+            }
+            _ => {}
+        }
+    }
+}
+
+impl<'a> Iterator for FileMerger<'a> {
+    type Item = String;
+    fn next(&mut self) -> Option<String> {
+        match self.heap.pop() {
+            Some(mut current) => {
+                match current.lines.next() {
+                    Some(Ok(next_line)) => {
+                        let ret = replace(&mut current.current_line, next_line);
+                        self.heap.push(current);
+                        return Some(ret);
```

Comment:
changed and rebased+force-pushed. I still find it tricky to spot implicit returns (and I've been using coffeescript at work for almost a year ;-) )